packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3614,9 +3643,145 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+  }
+
+  void _handleContextMenuOnScroll(ScrollNotification notification) {
+    if (_webContextMenuEnabled) {
+      return;
+    }
+    if (!_platformSupportsFadeOnScroll) {
+      _selectionOverlay?.updateForScroll();
+      return;
+    }
+    // When the scroll begins and the toolbar is visible, hide it
+    // until scrolling ends.
+    //
+    // The selection and renderEditable need to be visible within the current
+    // viewport for the toolbar to show when scrolling ends. If they are not
+    // then the toolbar is shown when they are scrolled back into view, unless
+    // invalidated by a change in TextEditingValue.
+    if (notification is ScrollStartNotification) {
+      if (_dataWhenToolbarShowScheduled != null) {
+        return;
+      }
+      final bool toolbarIsVisible = _selectionOverlay != null
+                                  && _selectionOverlay!.toolbarIsVisible
+                                  && !_selectionOverlay!.spellCheckToolbarIsVisible;
+      final Rect? selectionBounds = toolbarIsVisible ? renderEditable.getBoxesForSelection(_value.selection)
+                                           .map((TextBox box) => box.toRect())
+                                           .reduce((Rect result, Rect rect) => result.expandToInclude(rect)) : null;
+      _dataWhenToolbarShowScheduled = toolbarIsVisible ? (value: _value, selectionBounds: selectionBounds!) : null;
+      if (_dataWhenToolbarShowScheduled != null) {
+        _selectionOverlay?.hideToolbar();
+      }
+    } else if (notification is ScrollEndNotification) {
+      if (_dataWhenToolbarShowScheduled == null) {
+        return;
+      }
+      if (_dataWhenToolbarShowScheduled!.value != _value) {
+        // Value has changed so we should invalidate any toolbar scheduling.
+        _dataWhenToolbarShowScheduled = null;
+        _disposeScrollNotificationObserver();
+        return;
+      }
+
+      final Rect deviceRect = _calculateDeviceRect();
+      final bool selectionVisibleInEditable = renderEditable.selectionStartInViewport.value || renderEditable.selectionEndInViewport.value;
+      final Rect selectionBounds = MatrixUtils.transformRect(renderEditable.getTransformTo(null), _dataWhenToolbarShowScheduled!.selectionBounds);
+      final bool selectionOverlapsWithDeviceRect = !selectionBounds.hasNaN && deviceRect.overlaps(selectionBounds);
+
+      if (selectionVisibleInEditable && selectionOverlapsWithDeviceRect && (_selectionInViewport(selectionBounds) ?? true)) {
+        showToolbar();
+        _dataWhenToolbarShowScheduled = null;
+      }
+    }
+  }
+
+  bool? _selectionInViewport(Rect selectionBounds) {
+    RenderAbstractViewport? closestViewport = RenderAbstractViewport.maybeOf(renderEditable);
+    if (closestViewport == null) {
+      return null;
+    }
+    while (closestViewport != null) {
+      final Rect closestViewportBounds = MatrixUtils.transformRect(closestViewport.getTransformTo(null), closestViewport.paintBounds);
```

Comment:
transforming both rects to the view coordinates seems to be a bit of an overkill? Maybe compare the rects in each viewport's coordinates until it reaches the topmost viewport?

Reply:
Good point, i'll change that!